src/vmm/src/devices/virtio/iov_deque.rs

```diff
@@ -0,0 +1,412 @@
+            libc::MAP_SHARED | libc::MAP_FIXED,
+            raw_memfd,
+            0,
+        )?;
+
+        // SAFETY:
+        // * `buffer` is valid both for reads and writes (allocated with `libc::PROT_READ |
+        //   libc::PROT_WRITE`. `
+        // * `buffer` is aligned at `PAGE_SIZE`
+        // * `buffer` points to memory allocated with a single system call to `libc::mmap`
+        let iov = unsafe {
+            std::slice::from_raw_parts_mut(
+                buffer.cast(),
+                2 * PAGE_SIZE / std::mem::size_of::<libc::iovec>(),
+            )
+        };
+
+        Ok(Self {
+            iov,
+            _memfd: memfd,
+            head: 0,
+            tail: 0,
+        })
+    }
+
+    /// Returns the number of `iovec` objects currently in the [`IovDeque`]
+    #[inline(always)]
+    pub(crate) fn len(&self) -> usize {
+        self.tail - self.head
+    }
+
+    /// Returns `true` if the [`IovDeque`] is empty, `false` otherwise
+    #[inline(always)]
+    pub(crate) fn is_empty(&self) -> bool {
+        self.head == self.tail
+    }
+
+    /// Returns `true` if the [`IovDeque`] is full, `false` otherwise
+    fn is_full(&self) -> bool {
+        self.len() == usize::from(FIRECRACKER_MAX_QUEUE_SIZE)
+    }
+
+    /// Resets the queue, dropping all its elements.
+    #[inline(always)]
+    pub(crate) fn clear(&mut self) {
+        self.head = 0;
+        self.tail = 0;
+    }
+
+    /// Adds an `iovec` in the ring buffer.
+    ///
+    /// Returns an `IovDequeError::Full` error if the buffer is full.
+    pub(crate) fn push_back(&mut self, iov: iovec) {
+        // This should NEVER happen, since our ring buffer is as big as the maximum queue size.
+        // We also check for the sanity of the VirtIO queues, in queue.rs, which means that if we
+        // ever try to add something in a full ring buffer, there is an internal bug in the device
+        // emulation logic. Panic here because the device is hopelessly broken.
+        if self.is_full() {
+            panic!("The number of `iovec` objects is bigger than the available space");
+        }
+
+        self.iov[self.tail] = iov;
+        self.tail += 1;
+    }
+
+    /// Gets an `iovec` from the beginning of the buffer.
+    ///
+    /// Returns an `IovDequeError::Empty` error if the buffer is empty.
+    pub(crate) fn pop_front(&mut self) -> Option<iovec> {
+        if self.is_empty() {
+            return None;
+        }
+
+        let iov = self.iov[self.head];
+        self.head += 1;
+        if self.head > usize::from(FIRECRACKER_MAX_QUEUE_SIZE) {
+            self.head -= usize::from(FIRECRACKER_MAX_QUEUE_SIZE);
+            self.tail -= usize::from(FIRECRACKER_MAX_QUEUE_SIZE);
+        }
+
+        Some(iov)
+    }
+
+    /// Drop the first `iovec` objects that describe guest memory of `size` length.
+    ///
+    /// This will drop `iovec` buffers which describe a total memory size of up to `size`. This
```

Comment:
```suggestion
    /// This will drop `iovec` buffers which describe a total memory size of at least `size`. This
```
no? Because if `size == 1` and the first iovec contains 10 bytes, then it drop iovecs describing 10 bytes

Reply:
I changed this one. Now I'm dropping by number of descriptors rather than "total size of covered bytes"